File: project_euler_py/problem_039.py
"""If p is the perimeter of a right angle triangle with integral length sides,
{a,b,c}, there are exactly three solutions for p = 120.

{20,48,52}, {24,45,51}, {30,40,50}

For which value of p ≤ 1000, is the number of solutions maximised?
"""

p_s = [0 for x in range(1001)]

# For complicated math reasons, for a right triangle with a
# perimeter less than 1000, its hypotenuse (c) must be
# less than approximately 500.
for c in range(1, 500):
    for b in range(1, c):
        for a in range(1, b):
            p = a + b + c
            if p > 1000: break
            if (a*a + b*b == c*c):
                p_s[p] += 1

print(p_s)
print(p_s.index(max(p_s)))

# Answer: 840
# Time: 0m6.725s

File: project_euler_rb/problem_039.rb
# If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120.

# {20,48,52}, {24,45,51}, {30,40,50}

# For which value of p ≤ 1000, is the number of solutions maximised?

ps = Hash.new(0)

# For complicated math reasons, for a right triangle with a
# perimeter less than 1000, its hypotenuse (c) must be
# less than approximately 500.
limit = 500

1.upto(limit) do |c|
  1.upto(c) do |b|
    1.upto(b) do |a|
      if (a*a + b*b == c*c)
        next unless a+b+c < 1000
        ps[a+b+c] += 1
      end
    end
  end
end

puts ps.max_by{ |k,v| v }[0]
# Complete!
# Answer = 840
